[clangd] Add CodePatterns config option under Completion

Noustaa opened the pull request:
Allows enabling/disabling code patterns & snippets suggestion from completion.
This can be done either with YAML config or CLI with --code-patterns

You can refer to this discussion for more context: https://github.com/clangd/clangd/discussions/1867

Noustaa committed:
[clangd] Add CodePatterns config option under Completion

Allows enabling/disabling code patterns & snippets suggestion from
completion.
This can be done either with YAML config or CLI with --code-patterns

Noustaa commented:
Code autocomplete works just fine with CodePatterns::None but include autocomplete does not work anymore.
Gonna investigate when i have some time.

HighCommander4 commented:
> Code autocomplete works just fine with CodePatterns::None but include autocomplete does not work anymore. Gonna investigate when i have some time.

I guess the preprocessor completions are all `RK_Pattern` (added [here](https://searchfox.org/llvm/rev/2f976956e5ccef566418853015fb6b31257aa69f/clang/lib/Sema/SemaCodeComplete.cpp#10144-10150), and the `CodeCompletionResult` constructor which [only takes a `CodeCompletionString`](https://searchfox.org/llvm/rev/2f976956e5ccef566418853015fb6b31257aa69f/clang/include/clang/Sema/CodeCompleteConsumer.h#917) sets the kind to `RK_Pattern`).

Noustaa commented:
There is 2 different features involved here.
1) The fact that if i type "inc" it will suggest me both includes snippets
Like the below image the ones tagged LSP on the right side
![image](https://github.com/user-attachments/assets/a21e1db2-f057-4460-921e-d62d25efb794)

2) The fact that if i type "#include <str" it is supposed to search my library path to suggest "string" as autocompletion for headers.

In the feature i'm trying to implement, number 1 should be OFF if CodePatterns is set to None but number 2 should still be working.
Are these 2 features (num 1 & 2) coupled or can they be managed separately as the current state ?

Noustaa committed:
fixup! [clangd] Add CodePatterns config option under Completion

Fix the #include preprocessor not autocompleting when CodePatterns is
set to None

Noustaa commented:
> It looks like we can achieve this by allowing `RK_Pattern` completions if `Context.getKind() == CodeCompletionContext::CCC_IncludedFile`.

Great ! That worked :)
I updated the PR and also added a unit test for this specific case

HighCommander4 commented on clang-tools-extra/clangd/tool/ClangdMain.cpp:
```diff
@@ -267,6 +267,17 @@ opt<Config::HeaderInsertionPolicy> HeaderInsertion{
             "Never insert #include directives as part of code completion")),
 };
 
+opt<Config::CodePatternsPolicy> CodePatterns{
```
We've been moving towards having configuration specified only in the config file, not on the command line.

In the few cases where an option exists in both places (e.g. `--function-arg-placeholders`), it's because the command-line argument pre-dated the introduction of the config file system, and was kept for backward-compatibility.

Noustaa replied:
Done [1430d62](https://github.com/llvm/llvm-project/pull/137613/commits/1430d62625fe84ecea5ade72343031ffddbc7197)

HighCommander4 commented on clang-tools-extra/clangd/CodeComplete.cpp:
```diff
@@ -950,6 +951,11 @@ struct CompletionRecorder : public CodeCompleteConsumer {
+      if (Config::current().Completion.CodePatterns ==
+              Config::CodePatternsPolicy::None &&
+          Result.Kind == CodeCompletionResult::RK_Pattern &&
+          ContextKind != CodeCompletionContext::CCC_IncludedFile)
```
Please add a comment explaining that this carveout is for filename completions

Noustaa replied:
Done [1c5d401](https://github.com/llvm/llvm-project/pull/137613/commits/1c5d4019d56e9c023b777dbf8e335f4f13e037fa)

HighCommander4 reviewed (changes requested):
Thanks, the patch and tests look pretty good. I think we should leave out the command-line arg, but otherwise should be good to go.

Noustaa committed:
fixup! [clangd] Add CodePatterns config option under Completion

Remove the CLI option

Noustaa committed:
fixup! [clangd] Add CodePatterns config option under Completion

This last condition is here to allow user to get autocomplete suggestion
when inside an include expression

HighCommander4 committed:
Capitalize config values in comment

to match the actual values the parser accepts

HighCommander4 reviewed (approved):
Thanks, LGTM!

(I made one small tweak, to ConfigFragment.h, to refer to the config option values `All` and `None` with correct capitalization.)

HighCommander4 commented:
@Noustaa if you'd like to also submit a documentation patch for https://clangd.llvm.org/config.html (source at https://github.com/llvm/clangd-www/blob/main/config.md), that would be appreciated!

Noustaa commented:
> @Noustaa if you'd like to also submit a documentation patch for https://clangd.llvm.org/config.html (source at https://github.com/llvm/clangd-www/blob/main/config.md), that would be appreciated!

Sure, i'm gonna do it tomorrow 🙂

Noustaa commented:
@HighCommander4 Documentation done: https://github.com/llvm/clangd-www/pull/131
